src/math/mod.rs

```diff
@@ -29,6 +29,124 @@ pub fn canon_egcd(a: i64, b: i64, c: i64) -> Option<(i64, i64, i64)> {
     }
 }
 
+fn pos_mod(n: i64, m: i64) -> i64 {
+    if n < 0 {
+        n + m
+    } else {
+        n
+    }
+}
+
+fn mod_mul(a: i64, b: i64, m: i64) -> i64 {
+    pos_mod((a as i128 * b as i128 % m as i128) as i64, m)
+}
+
+/// Assuming m >= 2 and exp >= 0, finds base ^ exp % m in logarithmic time
+pub fn mod_exp(mut base: i64, mut exp: i64, m: i64) -> i64 {
+    assert!(m >= 2);
+    assert!(exp >= 0);
+    let mut ans = 1 % m;
+    base = base % m;
+    while exp > 0 {
+        if exp % 2 == 1 {
+            ans = mod_mul(ans, base, m);
+        }
+        base = mod_mul(base, base, m);
+        exp /= 2;
+    }
+    pos_mod(ans, m)
+}
+
+/// Assuming m >= 2, finds multiplicative inverse of n under modulus m
+pub fn mod_inv(n: i64, m: i64) -> i64 {
+    mod_exp(n, m - 2, m)
+}
+
+fn miller_test(n: i64, d: i64, r: i64, a: i64) -> bool {
+    let mut x = mod_exp(a, d, n);
+    if x == 1 || x == n - 1 {
+        return true;
+    }
+    for _ in 0..r {
+        x = mod_mul(x, x, n);
+        if x == n - 1 {
+            return true;
+        }
+    }
+    false
+}
+
+const BASES: [i64; 12] = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37];
+/// Assuming x >= 0, returns true if x is prime
+pub fn is_prime(n: i64) -> bool {
+    assert!(n >= 0);
+    if n <= 1 {
+        return false;
+    }
+    if n <= 3 {
+        return true;
+    }
+    let mut d = n - 1;
+    let mut r = 0;
+    while d % 2 == 0 {
+        d /= 2;
+        r += 1;
+    }
+    for base in BASES.iter() {
```

Comment:
You could write this as `BASES.iter().all(|&base| miller_test(n, d, r, base))`
Is the `base <= n-2` check important?

Reply:
I don't really know why it is the case that the number should be in the range (1, n - 1), but it is stated so here https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test#Example and here https://www.geeksforgeeks.org/primality-test-set-3-miller-rabin/ , so I would think it is required.

Reply:
I don't really know why it is the case that the number should be in the range (1, n - 1), but it is stated so here https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test#Example and here https://www.geeksforgeeks.org/primality-test-set-3-miller-rabin/ , so I would think it is required.

Reply:
Hm I guess the only values outside [2, n-2] would be 0 and +/- 1 (mod n). Since the test is about finding non-trivial square roots of 1, it seems the check is needed. You could still write it in the shortened way:
`BASES.iter().all(|&base| base > n - 2 && miller_test(n, d, r, base))`

Reply:
That's neat, although I believe you meant to put an `||` in there instead of `&&`